int/codec.ts

```diff
@@ -36,51 +36,31 @@ export const i32 = new NumCodec(4, "Int32");
 export const u64 = new NumCodec(8, "BigUint64");
 export const i64 = new NumCodec(8, "BigInt64");
 
+// https://github.com/unstoppablejs/unstoppablejs/blob/2082bf14d9836e52d282d89921df0d6f76db8eda/packages/scale-ts/src/codecs/i128.ts
 class X128 extends Codec<bigint> {
   constructor(signed: boolean) {
     super(
       () => {
         return 16;
       },
       (cursor, value) => {
-        if (value < 0) {
-          value = BigInt.asUintN(16 * 8, value);
-        }
-        encodePositiveBigIntInto(value, cursor.u8a, cursor.i, 16);
+        cursor.view.setBigInt64(cursor.i, value, true);
```

Comment:
I don't think that's correct
The documentation of `setBigInt64` mentions that if `value` is too large, then it will write `-9223372036854775808n`.

Reply:
:thinking:  I still think that the code is correct, I think that the docs are confusing, what happens when there is overflow is that it truncates the overflowing bytes, which in the example of the docs results on that particular value.

Although, I could be missing something here, could you please provide me with a value where the codec wouldn't work as expected :pray: ?

Reply:
Here is [an example](https://stackblitz.com/edit/typescript-vk9e6l?file=index.ts) that uses `scale-ts`, which has this exact same implementation.

Reply:
Ok, I guess that I should explain why this code is in fact correct:

First, let me clarify what [the docs mean by this](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView/setBigInt64#value):

> The value to set as a [BigInt](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt). The highest possible value that fits in a signed 64-bit integer is 2n ** (64n -1n) - 1n (9223372036854775807n). Upon overflow, it will be negative (-9223372036854775808n).

The maximum 64-bit **signed** int value is: `9223372036854775807n`, which (in big-endian) looks like this: `0x7fffffffffffffff`, which translated to binary means: one `0`, followed by sixty three `1`, awesome! Then, if we add `+1` to that huge number, then we would require another bit to represent it in complement two (because it should start with zero). However, we only have 64 bits at our disposal, so the zero overflows, which means that when adding one to `0x7fffffffffffffff`, the value becomes  `0x8000000000000000`, which translated to binary means: one `1` followed by sixty three `0`. That value happens to be the minimum 64-bit **signed** int value (`-9223372036854775808n`).

So, essentially what the docs are saying is that: if you try to put a very large number, then take into account that the overflowing bytes will be truncated. Which is exactly what we want!

Why?

Well, because when setting a 128-bit int value (signed or unsigned), if we want to encode it in little-endian that means that the "last" 8 bytes must go unaltered (truncated) as the first 8 bytes (but reversed), which if we think about it: that's exactly what `setBigInt64(cursor.i, value, true)` is doing.

And... As I was writing this, I did realize that the next line can actually be simplified as:

```ts
cursor.view.setBigInt64(cursor.i + 8, value >> 64n, true)
```
for both signed and unsigned :joy:. It's just when decoding that we must use different getters! Well, now I'm happy that I decided to explain this :sweat_smile:

Reply:
Testing in a browser console, the behavior lines up with what @josepot describes, and the behavior of the codec does seem to be correct from some testing.

Reply:
Also, I didn't mention it, but I did review [the spec](https://tc39.es/ecma262/multipage/structured-data.html#sec-dataview.prototype.setbigint64) and I tried to describe it in a more "human friendly" way, the most relevant parts are [here](https://tc39.es/ecma262/multipage/structured-data.html#sec-numerictorawbytes) and [here](https://tc39.es/ecma262/multipage/abstract-operations.html#sec-tobiguint64).